storage/namespace.go

```diff
@@ -0,0 +1,280 @@
+	shardID := n.shardSet.Shard(id)
+	shard, err := n.shardAt(shardID)
+	if err != nil {
+		return nil, err
+	}
+	return shard.ReadEncoded(ctx, id, start, end)
+}
+
+func (n *dbNamespace) FetchBlocks(
+	ctx context.Context,
+	shardID uint32,
+	id string,
+	starts []time.Time,
+) ([]FetchBlockResult, error) {
+	shard, err := n.shardAt(shardID)
+	if err != nil {
+		return nil, xerrors.NewInvalidParamsError(err)
+	}
+	return shard.FetchBlocks(ctx, id, starts), nil
+}
+
+func (n *dbNamespace) FetchBlocksMetadata(
+	ctx context.Context,
+	shardID uint32,
+	limit int64,
+	pageToken int64,
+	includeSizes bool,
+) ([]FetchBlocksMetadataResult, *int64, error) {
+	shard, err := n.shardAt(shardID)
+	if err != nil {
+		return nil, nil, xerrors.NewInvalidParamsError(err)
+	}
+	res, nextPageToken := shard.FetchBlocksMetadata(ctx, limit, pageToken, includeSizes)
+	return res, nextPageToken, nil
+}
+
+func (n *dbNamespace) Bootstrap(
+	bs bootstrap.Bootstrap,
+	writeStart time.Time,
+	cutover time.Time,
+) error {
+	n.Lock()
+	if n.bs == bootstrapped {
+		n.Unlock()
+		return nil
+	}
+	if n.bs == bootstrapping {
+		n.Unlock()
+		return errNamespaceIsBootstrapping
+	}
+	n.bs = bootstrapping
+	n.Unlock()
+
+	if !n.nopts.GetNeedsBootstrap() {
+		n.Lock()
+		n.bs = bootstrapped
+		n.Unlock()
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		err := shard.Bootstrap(bs, n.name, writeStart, cutover)
+		multiErr = multiErr.Add(err)
+	}
+
+	n.Lock()
+	n.bs = bootstrapped
+	n.Unlock()
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) Flush(
+	ctx context.Context,
+	blockStart time.Time,
+	pm persist.Manager,
+) error {
+	n.RLock()
+	if n.bs != bootstrapped {
+		n.RUnlock()
+		return errNamespaceNotBootstrapped
+	}
+	n.RUnlock()
+
+	if !n.nopts.GetNeedsFlush() {
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		// NB(xichen): we still want to proceed if a shard fails to flush its data.
+		// Probably want to emit a counter here, but for now just log it.
+		if err := shard.Flush(ctx, n.name, blockStart, pm); err != nil {
+			detailedErr := fmt.Errorf("shard %d failed to flush data: %v", shard.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) CleanupFileset(earliestToRetain time.Time) error {
+	if !n.nopts.GetNeedsFilesetCleanup() {
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		if err := shard.CleanupFileset(n.name, earliestToRetain); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) Truncate() (int64, error) {
+	var totalNumSeries int64
+
+	n.RLock()
+	shards := n.shardSet.Shards()
+	for _, shard := range shards {
+		totalNumSeries += n.shards[shard].NumSeries()
+	}
+	n.RUnlock()
+
+	n.initShards()
```

Comment:
I believe it actually might not be as easy as calling `initShards`.  If we do it this way then nothing will be returned to all the pools.  I think we need to go and manually cleanup each series and call close on all the blocks, etc.  We should probably audit the "expired" series case too and make sure it does this. 

We could potentially do this in a follow up diff if we want to?